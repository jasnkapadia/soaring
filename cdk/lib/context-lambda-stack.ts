/**
 * Lambda stack for contextual engine (geoIP and resource details)
 */

import * as cdk from '@aws-cdk/core';
import * as s3 from "@aws-cdk/aws-s3";
import * as iam from "@aws-cdk/aws-iam";

export class TargetStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    
  }
}
