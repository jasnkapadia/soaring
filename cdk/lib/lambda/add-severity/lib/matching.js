import yaml from "js-yaml";
import jsonpath from "jsonpath";
import * as matchers from "../matchers";

import fs from "fs";
const patternDefinitions = yaml.load(fs.readFileSync("./config/patterns.yml"));

function matchRegex(patternName, definition, data) {
  if (typeof definition?.pattern?.value !== "string") {
    throw new TypeError(
      `Pattern matcher value for ${patternName} must be a string.`
    );
  }

  const pattern = new RegExp(definition.pattern.value);
  return pattern.test(data);
}

function matchAnythingBut(patternName, definition, data) {
  return !matchRegex(patternName, definition, data);
}

function matchCustom(patternName, definition, data) {
  if (
    Object.prototype.hasOwnProperty.call(matchers, definition.pattern.value)
  ) {
    return matchers[definition.pattern.value](data);
  }

  throw new Error(
    `Custom pattern matcher ${definition.pattern.value} for ${patternName} does not exist.`
  );
}

function getMatcher(patternName, definition, data) {
  const patternMatchers = {
    matches: matchRegex,
    anything_but: matchAnythingBut,
    custom: matchCustom,
  };

  if (
    !Object.prototype.hasOwnProperty.call(
      patternMatchers,
      definition.pattern.type
    )
  ) {
    throw new Error(
      `Pattern matcher type ${definition.pattern.type} for ${patternName} is not valid.`
    );
  }

  return patternMatchers[definition.pattern.type](
    patternName,
    definition,
    data
  );
}

/** Logic for matching patterns */
export function matchPattern(event, patternName) {
  const definition = patternDefinitions.patterns[patternName];
  if (definition?.path === undefined)
    throw new TypeError(`Missing path for ${patternName}`);
  if (definition?.pattern === undefined)
    throw new TypeError(`Missing pattern for ${patternName}`);

  let data;
  try {
    data = jsonpath.value(event, definition.path);
  } catch (e) {
    console.log(`Skipping: Couldn't match event using ${patternName}`);
    return false;
  }

  const isMatching = getMatcher(patternName, definition, data);
  return isMatching;
}
