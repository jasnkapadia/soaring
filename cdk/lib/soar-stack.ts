/**
 * This file contains the SOAR stack
 */

import { Effect, PolicyStatement } from '@aws-cdk/aws-iam';
import * as lambda from '@aws-cdk/aws-lambda';
import * as sfn from '@aws-cdk/aws-stepfunctions';
import * as tasks from '@aws-cdk/aws-stepfunctions-tasks';
import * as cdk from '@aws-cdk/core';
import { Duration, Tags } from "@aws-cdk/core";
import * as path from "path";
import { Extract } from './soar/extract';


export class SoarStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Add Lambdas
    const macieJobLambda = new lambda.Function(this, "MacieJobLambda", {
      code: lambda.Code.fromAsset(path.join(__dirname, "lambda/macie-job")),
      runtime: lambda.Runtime.PYTHON_3_8,
      handler: "index.lambda_handler",
      memorySize: 512,
      timeout: Duration.seconds(15)
    });

    macieJobLambda.addToRolePolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      actions: [
        "macie2:CreateClassificationJob"
      ],
      resources: ["*"]
    }));

    const macieStatusLambda = new lambda.Function(this, "MacieStatusLambda", {
      code: lambda.Code.fromAsset(path.join(__dirname, "lambda/macie-status")),
      runtime: lambda.Runtime.PYTHON_3_8,
      handler: "index.lambda_handler",
      memorySize: 512,
      timeout: Duration.seconds(15)
    });

    macieStatusLambda.addToRolePolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      actions: [
        "macie2:DescribeClassificationJob"
      ],
      resources: ["*"]
    }));

    const macieFindingLambda = new lambda.Function(this, "MacieFindingLambda", {
      code: lambda.Code.fromAsset(path.join(__dirname, "lambda/macie-finding")),
      runtime: lambda.Runtime.PYTHON_3_8,
      handler: "index.lambda_handler",
      memorySize: 512,
      timeout: Duration.seconds(15)
    });

    macieFindingLambda.addToRolePolicy(new PolicyStatement({
      effect: Effect.ALLOW,
      actions: [
        "macie2:GetFindings"
      ],
      resources: ["*"]
    }));

    // const getIdentityLambda = new lambda.Function(this, "IdLambda", {
    //   code: lambda.Code.fromAsset(path.join(__dirname, "lambda/get-identity")),
    //   runtime: lambda.Runtime.NODEJS_12_X,
    //   handler: "index.exports.handler",
    //   memorySize: 512,
    //   timeout: Duration.seconds(15)
    // });

    // const findingLambda = new lambda.Function(this, "FindingLambda", {
    //   code: lambda.Code.fromAsset(path.join(__dirname, "lambda/finding")),
    //   runtime: lambda.Runtime.PYTHON_3_8,
    //   handler: "index.lambda_handler",
    //   memorySize: 512,
    // });

    // findingLambda.addToRolePolicy(new PolicyStatement({
    //   effect: Effect.ALLOW,
    //   actions: [
    //     "securityhub:BatchImportFindings"
    //   ],
    //   resources: ["arn:aws:securityhub:ap-southeast-2:659855141795:product/659855141795/default"]
    // }));



    // Configure steps
    const macieJob = new tasks.LambdaInvoke(this, "MacieJobStep", {
      "lambdaFunction": macieJobLambda,
      "retryOnServiceExceptions": false,
      "inputPath": "$",
      "outputPath": "$"
    });

    const macieStatus = new tasks.LambdaInvoke(this, "MacieStatusStep", {
      "lambdaFunction": macieStatusLambda,
      "retryOnServiceExceptions": false,
      "inputPath": "$.Payload",
      "outputPath": "$"
    });

    const macieFinding = new tasks.LambdaInvoke(this, "MacieFindingStep", {
      "lambdaFunction": macieFindingLambda,
      "retryOnServiceExceptions": false,
      "inputPath": "$.Payload",
      "outputPath": "$"
    });
    
    const waitForMacieJob = new sfn.Wait(this, "waitForMacieJob", {
      "time": sfn.WaitTime.duration(Duration.seconds(60))
    });
    waitForMacieJob.next(macieStatus)
    
    const checkMacieStatus = new sfn.Choice(this, "checkMacieStatus");

    // const getIdentity = new tasks.LambdaInvoke(this, "IdStep", {
    //   "lambdaFunction": getIdentityLambda,
    //   "retryOnServiceExceptions": false,
    //   "inputPath": "$",
    //   "outputPath": "$"
    // });

    // const finding = new tasks.LambdaInvoke(this, "FindingsStep", {
    //   "lambdaFunction": findingLambda,
    //   "retryOnServiceExceptions": false,
    //   "inputPath": "$",
    //   "outputPath": "$"
    // })


    // Configure step function defintion
    const sfnDefinition = sfn.Chain
      .start(macieJob)
      .next(macieStatus)
      .next(checkMacieStatus
        .when(sfn.Condition.stringEquals('$.Payload.macieJobs.jobStatus', 'COMPLETE'),
          macieFinding
          // .next(getIdentity)
          // .next(finding)
          )
        .otherwise(waitForMacieJob));

    // Set up rest of infrastructure
    const stateMachine = new sfn.StateMachine(this, "SoaringSoln-Macie", {
      "stateMachineName": "soar-stack-macie",
      "stateMachineType": sfn.StateMachineType.STANDARD,
      "definition": sfnDefinition
    });
    

    new Extract(this, "ExtractComponent", { sfn: stateMachine });
    Tags.of(this).add("OWNER", "team");
  }
}
