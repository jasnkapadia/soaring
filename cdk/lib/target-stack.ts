/**
 * This file contains the target stack for
 */

import * as cdk from '@aws-cdk/core';
import * as s3 from "@aws-cdk/aws-s3";
import * as iam from "@aws-cdk/aws-iam";

export class TargetStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // 1. Demo as a "canary" bucket which acts as an alarm when someone is trying to look for data e.g. from phishing/leaked credentials
    const bucketTargetCanary = new s3.Bucket(this, "CompanyConfidentialSecrets", {
      encryption: s3.BucketEncryption.S3_MANAGED,
      blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL
    });

    // People who own the secrets but should only never access them (as the bucket is a canary).
    // const executivesGroup = new iam.Group(this, "Executives");
    // bucketTargetCanary.grantReadWrite(executivesGroup);
  }
}
