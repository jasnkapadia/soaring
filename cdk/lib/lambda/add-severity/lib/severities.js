import yaml from "js-yaml";
import * as matching from "./matching";

import fs from "fs";
const config = yaml.load(fs.readFileSync("./config/config.yml"));
const severities = config.severity;

export function getSeverity(name, event, severity) {
  if (matching.matchPattern(event, name)) {
    return severities[severity];
  }

  return null;
}

export function getNamedSeverity(severity) {
  return (
    Object.keys(severities).find((key) => severity <= severities[key]) ||
    "CRITICAL"
  );
}

export function getDisplaySeverity(severity) {
  return {
    description: getNamedSeverity(severity),
    score: severity,
  };
}
