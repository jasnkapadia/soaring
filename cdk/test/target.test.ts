import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import * as Cdk from '../lib/target-stack';

// Snapshots the state of the stack (if it doesn't match update the tests)
it('matches the snapshot', () => {
  const app = new cdk.App();
  const stack = new Cdk.TargetStack(app, 'MyTestStack');
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
