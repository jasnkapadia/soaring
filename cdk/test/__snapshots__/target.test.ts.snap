// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches the snapshot 1`] = `
Object {
  "Resources": Object {
    "CompanyConfidentialSecretsCD61A5A6": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
